instrumentation/opentelemetry-instrumentation-boto3sqs/src/opentelemetry/instrumentation/boto3sqs/__init__.py

```diff
@@ -0,0 +1,428 @@
+        return [value.get("StringValue", None)]
+
+    def keys(self, carrier: CarrierT) -> List[str]:
+        return [
+            key[len(OPENTELEMETRY_ATTRIBUTE_IDENTIFIER) :]
+            if key.startswith(OPENTELEMETRY_ATTRIBUTE_IDENTIFIER)
+            else key
+            for key in carrier.keys()
+        ]
+
+
+class Boto3SQSSetter(Setter):
+    def set(self, carrier: CarrierT, key: str, value: str) -> None:
+        # This is a limitation defined by AWS for SQS MessageAttributes size
+        if len(carrier.items()) < 10:
+            carrier[f"{OPENTELEMETRY_ATTRIBUTE_IDENTIFIER}{key}"] = {
+                "StringValue": value,
+                "DataType": "String",
+            }
+        else:
+            logger.warning(
+                "Boto3 SQS instrumentation: cannot set context propagation on SQS/SNS message due to maximum amount of "
+                "MessageAttributes"
+            )
+
+
+boto3sqs_getter = Boto3SQSGetter()
+boto3sqs_setter = Boto3SQSSetter()
+
+
+# pylint: disable=attribute-defined-outside-init
+class Boto3SQSInstrumentor(BaseInstrumentor):
+    received_messages_spans: Dict[str, Span] = {}
+    current_span_related_to_token: Span = None
+    current_context_token = None
+
+    class ContextableList(list):
+        """
+        Since the classic way to process SQS messages is using a `for` loop, without a well defined scope like a
+        callback - we are doing something similar to the instrumentaiton of Kafka-python and instrumenting the
+        `__iter__` functions and the `__getitem__` functions to set the span context of the addressed message. Since
+        the return value from an `SQS.ReceiveMessage` returns a builtin list, we cannot wrap it and change all of the
+        calls for `list.__iter__` and `list.__getitem__` - therefore we use ContextableList. It is bound to the
+        received_messages_spans dict
+        """
+
+        def __getitem__(
+            self, *args: List[Any], **kwargs: Dict[str, Any]
+        ) -> Any:
+            retval = super(
+                Boto3SQSInstrumentor.ContextableList, self
+            ).__getitem__(*args, **kwargs)
+            if not isinstance(retval, dict):
+                return retval
+            receipt_handle = retval.get("ReceiptHandle", None)
+            if not receipt_handle:
+                return retval
+            started_span = Boto3SQSInstrumentor.received_messages_spans.get(
+                receipt_handle, None
+            )
+            if not started_span:
+                return retval
+            if Boto3SQSInstrumentor.current_context_token:
+                context.detach(Boto3SQSInstrumentor.current_context_token)
+            Boto3SQSInstrumentor.current_context_token = context.attach(
+                trace.set_span_in_context(started_span)
+            )
+            Boto3SQSInstrumentor.current_span_related_to_token = started_span
+            return retval
+
+        def __iter__(self) -> Generator:
```

Comment:
Just curious, why this? Do you want to use a generator for performance reasons?

Reply:
In general when it comes to SQS queues, unlike other libraries where you define a callback that handles each message, here we don't have an actual context (like a callback) that starts and ends for handling a single message. That means we can't know which message span is the currently active one.
This is why I change the response to be of this proxy object type, where every access to an item (via `__getitem__`) or just iteration over the response (via `__iter__`) sets the context.
When I implement `__iter__` calling `__getitem__`, I maintain the context management only in that `__getitem__` 😄